[Tabs] API review: Bottom navigation support

brianjmoore opened the pull request:
This API change adds support for bottom navigation bars to MDCTabBar using UIBarPositioning. The the tab bar's delegate can override -positionForBar: and return UIBarPositionBottom to configure the tab bar for use as bottom navigation. 

Usage example:
https://github.com/brianjmoore/material-components-ios/blob/inplace-bottomnav-v2/components/Tabs/examples/BottomNavigationBarExample.m

brianjmoore committed:
[Tabs] API changes to support bottom navigation

ajsecord reviewed (changes requested):
Any reason all of this configuration couldn't be replaced with UI_APPEARANCE properties on the views themselves?

ianegordon commented on components/Tabs/src/MDCTabBar.h:
```diff
@@ -152,3 +172,28 @@ IB_DESIGNABLE
+ The tab bar copies the configuration passed at initialization, so all configuration changes must
+ be made before creating the tab bar. 
+*/
+@interface MDCTabBarConfiguration : NSObject <NSCopying>
```
Can we mimic Apple and attach UIBarPositioningDelegate to MDCTabBarDelegate?
Doing so might eliminate the top/bottomConfiguration and allow us to move the remaining properties to MDCTabBar as IBInspectable @properties.

brianjmoore replied:
I really like this idea!

My only concern would be that it makes the default values for affected properties weird. For example, the default alignment for top tabs should be Leading but for bottom tabs should be Justified. Similar deal for the other two properties that are public on the configuration. 

What would [[MDCTabBar alloc] init].alignment return? What if a delegate is set?
A. Specifically Leading or Justified, chosen arbitrarily or based on whatever the delegate returns.
B. A new sentinel alignment case Default that specifies it's whatever is the default value for the context. This approach works for enums and objects, but would require changing BOOL properties into an enum.
C. Explicitly undefined - querying the alignment before setting it is not correct use of the API. Prone to misuse, but clearly simplest to implement.

brianjmoore replied:
Or I suppose
D: Change all the properties to methods that take a bar position.

ianegordon replied:
We can pick the most common case for the default.  I would guess Top.

Could we add alignment to MDCTabBarPositioning / MDCTabBarPositioningDelegate? (Which would inherit from UIBarPositioning and UIBarPositioningDelegate respectively)

This matches Apple's use of default, toolbars are at the top and tabbars are at the bottom.

brianjmoore commented:
@ajsecord I had a few reasons for wrapping this all up into the configuration

* To avoid the complexity for the user of having to set a dozen or so properties to fairly specific values to achieve a particular look. For an example of the breadth of properties that would need to be set, see https://github.com/brianjmoore/material-components-ios/blob/inplace-bottomnav/components/Tabs/src/MDCTabBar.m#L371. Ian's UIBarPosition idea would take care of this problem if we can resolve the default values problem.
* Some of the properties we need to achieve the bottom navigation style are pretty darn fiddly. It boils down to essentially what we have on MDCItemBarStyle: https://github.com/material-components/material-components-ios/blob/develop/components/Tabs/src/private/MDCItemBarStyle.h. In particular, I'm not sure we want to support customizing things like the title-image padding, ink color or maximum item width. I suppose they can always be deprecated and become no-ops in the future, though.

brianjmoore committed:
[Tabs] Update proposed API to use UIBarPositioning

brianjmoore committed:
[Tabs] Small comment update

brianjmoore commented:
Updated proposal to use UIBarPositioning, and also updated draft implementation at:
https://github.com/material-components/material-components-ios/compare/develop...brianjmoore:inplace-bottomnav-v2

Overall this does simplify the API a ton, but has some drawbacks:
* Setting alignment/itemAppearance/displaysUppercaseTitles the first time now has special meaning and effectively imposes an override of the position-based default value.
* sizeThatFits: also now secretly depends on the bar position, and thus the delegate. This imposes a not-entirely-obvious ordering dependency between calling sizeThatFits: and setting the delegate.
* The implementation of MDCTabBar is a bit more complex due to the new behaviors for default properties.

Please take another look.

ianegordon commented:
With the changes in your second branch, LGTM.
